### PR TITLE
add json output format to configure file

### PR DESCRIPTION
**Pull request by bruchar1**

### CODE REVIEW

**Comment by bruchar1** [comment]
@eli-schwartz Could you please review this PR?

File: mesonbuild/utils/universal.py [hunk 1]
```diff
@@ -1361,34 +1365,33 @@ def do_conf_file(src: str, dst: str, confdata: 'ConfigurationData',
...
 '''
 
-def dump_conf_header(ofilename: str, cdata: 'ConfigurationData', output_format: T.Literal['c', 'nasm']) -> None:
-    if output_format == 'c':
-        prelude = CONF_C_PRELUDE
-        prefix = '#'
-    else:
-        prelude = CONF_NASM_PRELUDE
-        prefix = '%'
+def _dump_c_header(ofile: TextIOWrapper, prelude: str, prefix: str, desc_prefix: str, desc_sep: str, desc_suffix: str, cdata: ConfigurationData) -> None:
```
Comment (eli-schwartz) [inline mesonbuild/utils/universal.py:1368]: I think this would be a lot simpler if you just handled the output_format in dump_c_header instead of calculating it in dump_conf_header just to pass each bit of data as parameters to another function.

File: docs/markdown/snippets/json_output_format.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+## Added 'json' output_format to configure_file()
+
+When no input file is specified, `configure_file()` can now
```
Comment (eli-schwartz) [inline docs/markdown/snippets/json_output_format.md:3]:
```suggestion
When no input file is specified, [[configure_file]] can now
```

This makes a documentation hyperlink.

File: docs/markdown/snippets/json_output_format.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+generate a `json` file from given `cfg_data`.
```
Comment (eli-schwartz) [inline docs/markdown/snippets/json_output_format.md:4]:
```suggestion
generate a `json` file from the given [[@cfg_data]]. 
```

File: test cases/common/262 configure file output format/compare.py [hunk 4]
```diff
@@ -0,0 +1,5 @@
+from pathlib import Path
+import sys
+
+are_equals = Path(sys.argv[1]).read_text(encoding='utf-8') == Path(sys.argv[2]).read_text(encoding='utf-8')
+sys.exit(0 if are_equals else 1)
```
Comment (eli-schwartz) [inline test cases/common/262 configure file output format/compare.py:5]:
```python
with open(sys.argv[1], 'rb') as f, open(sys.argv[2], 'rb') as g:
    if f.read() != g.read():
        sys.exit('contents are not equal')
```

- there's no real reason to open this as unicode text, since what we are checking is that meson produces the same file, but we don't actually care about interpreting what's in it
- the sys.exit seems complicated and hard to read compared to exiting with a message in an if block
- it is not clear to me what benefit pathlib brings here as no path manipulation is used and even the method names are unnecessarily longer IMO

File: test cases/common/262 configure file output format/meson.build [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+)
+
+config_json = configure_file(
+    configuration: data,
+    output_format: 'json',
+    output: 'config.json'
+)
+
+pymod = import('python')
+py = pymod.find_installation('python3')
```
Comment (eli-schwartz) [inline test cases/common/262 configure file output format/meson.build:30]: Why not `find_program('python3')`? Note that I *fixed* tons of test cases to use find_program instead, in #10361.

File: test cases/common/262 configure file output format/compare.py (hunk 4 above)
Comment (bruchar1) [reply test cases/common/262 configure file output format/compare.py:5]: IIRC, I did this in text mode instead of using filecmp, because of line endings that are different on Windows. By opening the file in text mode, Python uses universal newlines, and the files compare equally even if the reference file is from a different platform.

Comment (eli-schwartz) [reply test cases/common/262 configure file output format/compare.py:5]: Huh. I did not think of that.

I think this just matters for C/nasm, right? I wonder if it makes sense to guarantee consistency when generating those files by opening them with `newline=''`

Comment (bruchar1) [reply test cases/common/262 configure file output format/compare.py]: I think it is easier / simpler to keep the test in text mode for now. After all, this is just a script for a test. I rewrote it like you proposed.

**Comment by bruchar1** [comment]
@eli-schwartz Its all green now

File: mesonbuild/utils/universal.py [hunk 6]
```diff
@@ -1361,34 +1365,43 @@ def do_conf_file(src: str, dst: str, confdata: 'ConfigurationData',
 
 '''
 
-def dump_conf_header(ofilename: str, cdata: 'ConfigurationData', output_format: T.Literal['c', 'nasm']) -> None:
-    if output_format == 'c':
-        prelude = CONF_C_PRELUDE
-        prefix = '#'
-    else:
+def _dump_c_header(ofile: TextIOWrapper, output_format: str, cdata: ConfigurationData) -> None:
```
Comment (eli-schwartz) [inline mesonbuild/utils/universal.py:1368]: Apparently I missed mentioning this before... We usually use T.TextIO here, which is also shorter:
```diff
-TextIOWrapper
+T.TextIO
```

File: mesonbuild/utils/universal.py [hunk 7]
```diff
@@ -1361,34 +1365,43 @@ def do_conf_file(src: str, dst: str, confdata: 'ConfigurationData',
...
+    format_desc: T.Callable[[str], str]
+    if output_format == 'nasm':
```
Comment (eli-schwartz) [inline mesonbuild/utils/universal.py:1370]: Why is the order switched from the initial order, which was alphabetic:
```python
    if output_format == 'c':
        ...
    else:
```

File: mesonbuild/utils/universal.py [hunk 8]
```diff
@@ -1361,34 +1365,43 @@ def do_conf_file(src: str, dst: str, confdata: 'ConfigurationData',
...
         prelude = CONF_NASM_PRELUDE
         prefix = '%'
+        format_desc = lambda desc: '; ' + '\n; '.join(desc.splitlines()) + '\n'
+    else:  # c
+        prelude = CONF_C_PRELUDE
+        prefix = '#'
+        format_desc = lambda desc: '/* ' + '\n'.join(desc.splitlines()) + ' */\n'
```
Comment (eli-schwartz) [inline mesonbuild/utils/universal.py:1377]: What's the purpose of the change from:
```
'%s */\n' % desc
```
to
```
'\n'.join(desc.splitlines())
```

We're splitting lines then joining them again as lines?

File: mesonbuild/utils/universal.py [hunk 9]
```diff
@@ -2437,8 +2450,8 @@ def pickle_load(filename: str, object_name: str, object_type: T.Type[_PL]) -> _P
     else:
         version = T.cast('_EnvPickleLoadable', obj).environment.coredata.version
 
+    from ..coredata import MesonVersionMismatchException, major_versions_differ
     from ..coredata import version as coredata_version
-    from ..coredata import major_versions_differ, MesonVersionMismatchException
```
Comment (eli-schwartz) [inline mesonbuild/utils/universal.py:2441]: Extraneous change?

File: test cases/common/262 configure file output format/meson.build [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    configuration: data,
+    output_format: 'json',
+    output: 'config.json'
+)
+
+py = find_program('python3')
+
+test('c_output', py, args: [meson.source_root() / 'compare.py', meson.source_root() / 'expected/config.h', config_h])
+test('nasm_output', py, args: [meson.source_root() / 'compare.py', meson.source_root() / 'expected/config.nasm', config_nasm])
+test('json_output', py, args: [meson.source_root() / 'compare.py', meson.source_root() / 'expected/config.json', config_json])
```
Comment (eli-schwartz) [inline test cases/common/262 configure file output format/meson.build:33]: You can use `files('compare.py')` etc here instead of source_root() which is deprecated anyways.

File: mesonbuild/utils/universal.py (hunk 8 above)
Comment (bruchar1) [reply mesonbuild/utils/universal.py]: oups. That's what happen when you refactor a refactor...

File: docs/markdown/snippets/json_output_format.md [hunk 11]
```diff
@@ -0,0 +1,5 @@
+## Added 'json' output_format to configure_file()
+
+When no input file is specified, [[configure_file()]] can now
```
Comment (eli-schwartz) [inline docs/markdown/snippets/json_output_format.md:3]: My suggestion removed the `()` when using the `[[]]` link formatter. I'm not sure it matters for the output generation but we usually don't include that...

File: test cases/common/262 configure file output format/compare.py [hunk 12]
```diff
@@ -0,0 +1,5 @@
+import sys
+
+with open(sys.argv[1], 'r', encoding='utf-8') as f, open(sys.argv[2], 'r', encoding='utf-8') as g:
+    if f.read() != g.read():
+        sys.exit('contents are not equal')
```
Comment (Volker-Weissmann) [inline test cases/common/262 configure file output format/compare.py:5]: This file can be written shorter:
```
import sys
import filecmp

assert filecmp.cmp(sys.argv[1], sys.argv[2], shallow=False)
```

File: test cases/common/262 configure file output format/meson.build [hunk 13]
```diff
@@ -0,0 +1,37 @@
+project('configure file output format')
+
+data = configuration_data()
+data.set_quoted('str', 'hello world!', description: '''This is
```
Comment (Volker-Weissmann) [inline test cases/common/262 configure file output format/meson.build:4]: Replace this line with
```
data.set_quoted('quoted', 'hello world!', description: '''This is
a multiline
description''')
data.set('unquoted', 'float')
```
to increase the test coverage. (Then update expected/*)

**Review by Volker-Weissmann — COMMENTED** [review]
I just realized that `set_quoted` is broken (independent of this PR), since it escapes ", but does not escape other characters like newline or the backslash. I will fix this in another PR.

Otherwise, LGTM.

File: test cases/common/262 configure file output format/compare.py (hunk 12 above)
Comment (bruchar1) [reply test cases/common/262 configure file output format/compare.py:5]: Unfortunately, this doesn't work because of line endings on Windows.

**Comment by eli-schwartz** [comment]
> I just realized that `set_quoted` is broken (independent of this PR), since it escapes ", but does not escape other characters like newline or the backslash. I will fix this in another PR.

I agree it's broken, and it was fixed back in 2019 by @Jehan. Then it got reverted because GLib is depending on the lack of escaping. (???)

The original fix is supposed to be in #6267.

**Comment by Jehan** [comment]
> The "new" fix is supposed to be in https://github.com/mesonbuild/meson/pull/6267.

Yeah I need to look again at #6267 to fix the unit tests (if I recall, my fix was uncovering broken code paths). It's like I find time to look at this patch once every year or something. 😅

Anyway I'll try to make some time soonish.

**Review by eli-schwartz — APPROVED** [review]
